### ISSUE TITLE
Psalm erroneously warns if nullable return types are narrowed to non-nullable

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#348]
This may have started happening after fixing https://github.com/vimeo/psalm/issues/288.

PHP allows this code to be executed, and returning a non-nullable string instead of a nullable string is valid.
Note that the opposite is not allowed - Non-nullable can't be overridden with nullable.

- Other edge cases: array is a narrower form of iterable, so array return type can override iterable return type, and iterable param can override array param.

Observed: MethodSignatureMismatch
Expected: No warnings

```php
<?php

interface I1 {
    public function test() : ?string;
    public function testIterable() : iterable;
}

class A1 implements I1 {
    public function test() : string {
        return "value";
    }
    public function testIterable() : array {
        return [];
    }
}
$a = new A1();
var_export($a->test());
var_export($a->testIterable());
```

**Comment by muglug** [comment]
Sorry, just fixed the first part

**Comment by muglug** [comment]
Ah, the merest sprinkling of return type covariance/param type contravariance
